### PR TITLE
Heedls 370 reset password logic

### PR DESCRIPTION
**Pull request by SteveJacksonSoft**
The backend for actually changing the password and the success page with the link to the login page.

Note: the latest commit changed PasswordDataServiceTests.cs and TempDataExtension LF -> CRLF (commit hash: 95f700943be571a9a227e4d40dc278b353b719ec) - you can ignore this commit when reviewing. 

Tested:
- Can't sneak in
  - Post form properties manually 
    - Not tested - assuming antiforgery token stops this
  - Temp data is cleared
    - Error occurs
    - Success
  - Post form with expired code
- Updates passwords for all the relevant admin and delegates
- Login link works
- Can't reach reset password form again with same code.

Success page:
SuccessPage

### CODE REVIEW

**Review by AlexJacksonDS — APPROVED** [review]
Overall looks fine, just one minor change for a unit test and a question about why the unit test is set up like it is.

File: DigitalLearningSolutions.Data/Services/PasswordResetService.cs [hunk 1]
```diff
@@ -75,6 +72,16 @@ public void GenerateAndSendPasswordResetLink(string emailAddress, string baseUrl
             emailService.SendEmail(resetPasswordEmail);
```
Comment (stellake) [inline DigitalLearningSolutions.Data/Services/PasswordResetService.cs:72]: I think FirstName can technically be null in the db, so we should handle it in GeneratePasswordResetEmail :)

File: DigitalLearningSolutions.Data.Tests/DataServices/PasswordResetDataServiceTests.cs [hunk 2]
```diff
@@ -4,24 +4,32 @@ namespace DigitalLearningSolutions.Data.Tests.DataServices
...
         [DatapointSource]
-        public UserType[] UserTypes = {UserType.AdminUser, UserType.DelegateUser};
+        public UserType[] UserTypes = { UserType.AdminUser, UserType.DelegateUser };
 
-        private readonly PasswordResetDataService service =
-            new PasswordResetDataService(
-                ServiceTestHelper.GetDatabaseConnection(),
-                new NullLogger<PasswordResetDataService>());
+        private PasswordResetDataService service = null!;
+        private SqlConnection connection = null!;
```
[the remark this answers is not in the document]
Comment (SteveJacksonSoft) [reply DigitalLearningSolutions.Data.Tests/DataServices/PasswordResetDataServiceTests.cs:25]: Yep! My experience with warnings is that unless you're fairly ruthless with them, they get to a critical mass where build warnings are completely ignored for the rest of the project. Open to having my opinion changed though!

[the remark this answers is not in the document]
Comment (stellake) [reply DigitalLearningSolutions.Data.Tests/DataServices/PasswordResetDataServiceTests.cs:25]: Oh, I actually think it's really good - just wanted to double check that that's the reason 👍

File: DigitalLearningSolutions.Data.Tests/DataServices/PasswordDataServiceTests.cs [hunk 3]
```diff
@@ -40,5 +40,75 @@ public void Set_password_by_candidate_number_should_set_password_correctly()
...
+
+            // Given
+            var existingAdminUser = UserTestHelper.GetDefaultAdminUser();
+            var newPasswordHash = "I haven't used this password before!";
+
+            // When
+            await this.passwordDataService.SetPasswordByEmailAsync(existingAdminUser.EmailAddress!, newPasswordHash);
+
+            // Then
+            this.userDataService.GetAdminUserById(existingAdminUser.Id)?.Password.Should()
```
[the remark this answers is not in the document]
Comment (SteveJacksonSoft) [reply DigitalLearningSolutions.Data.Tests/DataServices/PasswordDataServiceTests.cs]: Yep, though it should possibly go in .editorconfig if you don't want redundant `this` s everywhere.

File: DigitalLearningSolutions.Data.Tests/DataServices/PasswordResetDataServiceTests.cs [hunk 4]
```diff
@@ -90,24 +111,153 @@ public async Task Does_Not_Match_Reset_Passwords_If_No_Reset_Password_With_Given
...
+
+            // When
+            await service.RemoveResetPasswordAsync(resetPasswordId);
+
+            // Then
+            var matchingResetPasswords = connection.Query<ResetPassword>
+            (
+                "SELECT * FROM ResetPassword WHERE ID = @ResetPasswordId",
+                new { ResetPasswordId = resetPasswordId }
+            );
```
Comment (stellake) [inline DigitalLearningSolutions.Data.Tests/DataServices/PasswordResetDataServiceTests.cs:193]: There are some test helper under "Helpers" folder that have the queries that we need only for tests - maybe worth separating this query into a helper too for consistency and better readability? (or just into a private function, if you prefer)

File: DigitalLearningSolutions.Data.Tests/DataServices/PasswordDataServiceTests.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (stellake) [reply DigitalLearningSolutions.Data.Tests/DataServices/PasswordDataServiceTests.cs]: Are you happy to add it? 🙏

[the remark this answers is not in the document]
Comment (SteveJacksonSoft) [reply DigitalLearningSolutions.Data.Tests/DataServices/PasswordDataServiceTests.cs]: Think that's it sorted!

File: DigitalLearningSolutions.Data.Tests/DataServices/PasswordResetDataServiceTests.cs [hunk 5]
```diff
@@ -90,24 +110,145 @@ public async Task Does_Not_Match_Reset_Passwords_If_No_Reset_Password_With_Given
...
+
+            foreach (var user in userList.Skip(1))
+            {
+                await GivenUserHasResetPasswordIdAsync(resetPasswordId, user);
+            }
+
+            return resetPasswordId;
+        }
+
+        private async Task<int> GetResetPasswordIdByHashAsync(string hash)
```
Comment (stellake) [inline DigitalLearningSolutions.Data.Tests/DataServices/PasswordResetDataServiceTests.cs:235]: Given that we have the test helpers now, should we move these queries across too?

File: DigitalLearningSolutions.Data/Helpers/ResetPasswordHelpers.cs [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+        public static IEnumerable<int> GetDistinctResetPasswordIds(this (AdminUser?, List<DelegateUser>) userAccounts)
+        {
+            var (adminUserIfAny, delegateUsers) = userAccounts;
+
+            return delegateUsers
+                .Select(du => (User)du)
+                .Concat(new[] { adminUserIfAny })
+                .Select(u => u?.ResetPasswordId ?? (int?) null)
+                .Where(rPId => rPId.HasValue)
+                .Select(rPId => rPId.Value)
```
Comment (stellake) [inline DigitalLearningSolutions.Data/Helpers/ResetPasswordHelpers.cs:27]: Can we silence this warning with "!"

**Review by stellake — APPROVED** [review]
Looks good! Minor comment about a warning & a location of specific test queries, but no need for rereview 👍

**Pull request merged**